### DISCUSSION TITLE
[Proposal]: override covariant support generic type parameter

### DISCUSSION THREAD

**Discussion by ahdung** [dotnet/csharplang#6439] [Language Ideas] [answered]
I want this:
```c#
// base member:
public virtual List<object> Foo() { ... }

// overriden member:
public override List<whatever> Foo() { ... }
```

**Comment by HaloFour** [comment, marked as answer]
This is not typesafe, for the same reason `List<T>` can't be variant in general.  Anyone accessing the base member would see `List<object>` and could attempt to add values to it that are not `whatever`, which would lead to runtime errors.

This is already permitted by covariant returns in C# where the return type of the overriding member is properly covariant, such as in the case of `IEnumerable<object>`.

**Comment by zms9110750** [comment]
```csharp
string[] s = new string[1];
object[] o = s;
o[0] = 100;
```
This code will not make mistakes.
 but it should not be allowed.

**Reply by HaloFour** [reply]
I think quite a few folks would agree that array variance was a mistake, but unfortunately it's been there since .NET and C# 1.0 and at this point it is very impractical to consider removing due to representing a massive breaking change.  At best analyzers could be used to detect and warn on unsafe usage.

See: https://github.com/dotnet/runtime/issues/23393

**Reply by CyrusNajmabadi** [reply]
Yes. Arrays predate generics and have this unfortunate behavior. We very intentionally did but carry it over to generics.
